File: citizen_connect_python/app/services/auth_service.py
from datetime import datetime, timedelta, timezone
from jose import jwt
from passlib.context import CryptContext
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.config import settings
from app.models.identity import Citizen, InternalUser, UserRole
from app.repositories.citizen_repository import CitizenRepository
from app.schemas.auth import (
    CitizenRegisterSchema, LoginSchema, AuthResponseSchema, UserInfoSchema
)
import uuid

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


class AuthService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.citizen_repo = CitizenRepository(db)

    async def register_citizen(
        self, dto: CitizenRegisterSchema
    ) -> AuthResponseSchema:
        if await self.citizen_repo.phone_exists(dto.phone):
            raise ValueError("Phone number already registered.")

        citizen = Citizen(
            id=uuid.uuid4(),
            full_name=dto.full_name,
            phone=dto.phone,
            email=dto.email,
            password_hash=pwd_context.hash(dto.password),
            block_id=dto.block_id,
            is_verified=False,
            is_active=True,
            created_at=datetime.now(timezone.utc)
        )
        await self.citizen_repo.create(citizen)
        return self._generate_citizen_token(citizen)

    async def login_citizen(self, dto: LoginSchema) -> AuthResponseSchema:
        citizen = await self.citizen_repo.get_by_phone(dto.identifier)
        if not citizen or not pwd_context.verify(dto.password, citizen.password_hash):
            raise PermissionError("Invalid credentials.")
        if not citizen.is_active:
            raise PermissionError("Account is deactivated.")

        citizen.last_login_at = datetime.now(timezone.utc)
        await self.citizen_repo.update(citizen)
        return self._generate_citizen_token(citizen)

    async def login_internal_user(self, dto: LoginSchema) -> AuthResponseSchema:
        result = await self.db.execute(
            select(InternalUser).where(
                InternalUser.email == dto.identifier,
                InternalUser.is_active == True
            )
        )
        user = result.scalar_one_or_none()
        if not user or not pwd_context.verify(dto.password, user.password_hash):
            raise PermissionError("Invalid credentials.")

        user.last_login_at = datetime.now(timezone.utc)
        await self.db.flush()
        return self._generate_internal_token(user)

    def _generate_citizen_token(self, citizen: Citizen) -> AuthResponseSchema:
        expiry = datetime.now(timezone.utc) + timedelta(
            hours=settings.jwt_expiry_hours
        )
        payload = {
            "sub": str(citizen.id),
            "name": citizen.full_name,
            "phone": citizen.phone,
            "user_type": "citizen",
            "block_id": citizen.block_id,
            "exp": expiry
        }
        token = jwt.encode(
            payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm
        )
        return AuthResponseSchema(
            token=token,
            refresh_token=str(uuid.uuid4()),
            expires_at=expiry.isoformat(),
            user=UserInfoSchema(
                id=str(citizen.id),
                full_name=citizen.full_name,
                phone=citizen.phone,
                email=citizen.email,
                user_type="citizen"
            )
        )

    def _generate_internal_token(self, user: InternalUser) -> AuthResponseSchema:
        expiry = datetime.now(timezone.utc) + timedelta(
            hours=settings.jwt_expiry_hours
        )
        role_name = UserRole(user.role).name
        payload = {
            "sub": str(user.id),
            "name": user.full_name,
            "email": user.email,
            "user_type": "internal",
            "role": role_name,
            "exp": expiry
        }
        token = jwt.encode(
            payload, settings.jwt_secret_key, algorithm=settings.jwt_algorithm
        )
        return AuthResponseSchema(
            token=token,
            refresh_token=str(uuid.uuid4()),
            expires_at=expiry.isoformat(),
            user=UserInfoSchema(
                id=str(user.id),
                full_name=user.full_name,
                email=user.email,
                user_type="internal",
                role=role_name
            )
        )

File: Application/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Core.DTOs.Auth;
using Core.Entities.Identity;
using Core.Interfaces.Repositories;
using Core.Interfaces.Services;
using Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;

namespace Application.Services
{
    public class AuthService : IAuthService
    {
        private readonly ICitizenRepository _citizenRepo;
        private readonly AppDbContext _db;
        private readonly IConfiguration _config;

        public AuthService(ICitizenRepository citizenRepo, AppDbContext db, IConfiguration config)
        {
            _citizenRepo = citizenRepo;
            _db = db;
            _config = config;
        }

        public async Task<AuthResponseDto> RegisterCitizenAsync(CitizenRegisterDto dto)
        {
            if (await _citizenRepo.PhoneExistsAsync(dto.Phone))
                throw new InvalidOperationException("Phone number already registered.");

            var citizen = new Citizen
            {
                FullName = dto.FullName,
                Phone = dto.Phone,
                Email = dto.Email,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password),
                BlockId = dto.BlockId,
                IsVerified = false
            };

            await _citizenRepo.CreateAsync(citizen);
            return GenerateCitizenToken(citizen);
        }

        public async Task<AuthResponseDto> LoginCitizenAsync(LoginDto dto)
        {
            var citizen = await _citizenRepo.GetByPhoneAsync(dto.Identifier);
            if (citizen == null || !BCrypt.Net.BCrypt.Verify(dto.Password, citizen.PasswordHash))
                throw new UnauthorizedAccessException("Invalid credentials.");

            if (!citizen.IsActive)
                throw new UnauthorizedAccessException("Account is deactivated.");

            citizen.LastLoginAt = DateTime.UtcNow;
            await _citizenRepo.UpdateAsync(citizen);

            return GenerateCitizenToken(citizen);
        }

        public async Task<AuthResponseDto> LoginInternalUserAsync(LoginDto dto)
        {
            var user = await _db.InternalUsers
                .FirstOrDefaultAsync(x => x.Email == dto.Identifier && x.IsActive);

            if (user == null || !BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
                throw new UnauthorizedAccessException("Invalid credentials.");

            user.LastLoginAt = DateTime.UtcNow;
            _db.InternalUsers.Update(user);
            await _db.SaveChangesAsync();

            return GenerateInternalUserToken(user);
        }

        private AuthResponseDto GenerateCitizenToken(Citizen citizen)
        {
            var claims = new[]
            {
            new Claim(ClaimTypes.NameIdentifier, citizen.Id.ToString()),
            new Claim(ClaimTypes.Name, citizen.FullName),
            new Claim(ClaimTypes.MobilePhone, citizen.Phone),
            new Claim("user_type", "citizen"),
            new Claim("block_id", citizen.BlockId.ToString())
        };
            return BuildAuthResponse(claims, new UserInfoDto
            {
                Id = citizen.Id.ToString(),
                FullName = citizen.FullName,
                Phone = citizen.Phone,
                Email = citizen.Email,
                UserType = "citizen"
            });
        }

        private AuthResponseDto GenerateInternalUserToken(InternalUser user)
        {
            var claims = new[]
            {
            new Claim(ClaimTypes.NameIdentifier, user.Id.ToString()),
            new Claim(ClaimTypes.Name, user.FullName),
            new Claim(ClaimTypes.Email, user.Email),
            new Claim(ClaimTypes.Role, user.Role.ToString()),
            new Claim("user_type", "internal"),
        };
            return BuildAuthResponse(claims, new UserInfoDto
            {
                Id = user.Id.ToString(),
                FullName = user.FullName,
                Email = user.Email,
                UserType = "internal",
                Role = user.Role.ToString()
            });
        }

        private AuthResponseDto BuildAuthResponse(Claim[] claims, UserInfoDto userInfo)
        {
            var key = new SymmetricSecurityKey(
                Encoding.UTF8.GetBytes(_config["Jwt:Key"]!));
            var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
            var expiry = DateTime.UtcNow.AddHours(
                int.Parse(_config["Jwt:ExpiryHours"] ?? "24"));

            var token = new JwtSecurityToken(
                issuer: _config["Jwt:Issuer"],
                audience: _config["Jwt:Audience"],
                claims: claims,
                expires: expiry,
                signingCredentials: creds
            );

            return new AuthResponseDto
            {
                Token = new JwtSecurityTokenHandler().WriteToken(token),
                RefreshToken = Guid.NewGuid().ToString(),
                ExpiresAt = expiry,
                User = userInfo
            };
        }
    }
}
